framework/src/modules/interoperability/utils.ts

```diff
@@ -593,3 +618,347 @@ export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
+			throw new Error('The total BFT weight of all active validators is not valid.');
+		}
+
+		await chainValidatorsStore.setWithSchema(
+			chainValidators.storeKey,
+			chainValidators.storeValue,
+			chainValidatorsSchema,
+		);
+	}
+
+	const chainDataStoreKeySet = new dataStructures.BufferSet();
+	const chainDataStore = getStore(id, STORE_PREFIX_CHAIN_DATA);
+	let isAnotherSidechainAccount = 0;
+	for (const chainData of genesisStore.chainDataSubstore) {
+		const chainDataStoreKey = chainData.storeKey;
+		if (chainDataStoreKeySet.has(chainDataStoreKey)) {
+			throw new Error(`Chain data store key ${chainDataStoreKey.toString('hex')} is duplicated.`);
+		}
+		chainDataStoreKeySet.add(chainDataStoreKey);
+
+		const chainAccountStatus = chainData.storeValue.status;
+		if (chainAccountStatus === CHAIN_TERMINATED) {
+			if (outboxRootStoreKeySet.has(chainDataStoreKey)) {
+				throw new Error('Outbox root store cannot have entry for a terminated chain account.');
+			}
+			if (
+				!channelDataStoreKeySet.has(chainDataStoreKey) ||
+				!chainValidatorsStoreKeySet.has(chainDataStoreKey)
+			) {
+				throw new Error(
+					`Chain data store key ${chainDataStoreKey.toString(
+						'hex',
+					)} missing in some or all of channel data and chain validators stores.`,
+				);
+			}
+		}
+		if (
+			!outboxRootStoreKeySet.has(chainDataStoreKey) ||
+			!channelDataStoreKeySet.has(chainDataStoreKey) ||
+			!chainValidatorsStoreKeySet.has(chainDataStoreKey)
+		) {
+			throw new Error(
+				`Chain data store key ${chainDataStoreKey.toString(
+					'hex',
+				)} missing in some or all of outbox root, channel data and chain validators stores.`,
+			);
+		}
+
+		if (
+			!(
+				chainDataStoreKey.equals(ownChainAccount.id) ||
+				chainDataStoreKey.equals(MAINCHAIN_ID_BUFFER)
+			)
+		) {
+			isAnotherSidechainAccount = 1;
+		}
+
+		await chainDataStore.setWithSchema(
+			chainData.storeKey,
+			chainData.storeValue,
+			chainAccountSchema,
+		);
+	}
+
+	if (
+		isAnotherSidechainAccount &&
+		!(chainDataStoreKeySet.has(MAINCHAIN_ID_BUFFER) && chainDataStoreKeySet.has(ownChainAccount.id))
+	) {
+		throw new Error(
+			'If a chain account for another sidechain is present, then a chain account for the mainchain must be present, as well as the own chain account.',
+		);
+	}
+
+	for (const storeKey of outboxRootStoreKeySet) {
+		if (!chainDataStoreKeySet.has(storeKey)) {
+			throw new Error(
+				`Outbox root store key ${storeKey.toString('hex')} is missing in chain data store.`,
+			);
+		}
+	}
+
+	for (const storeKey of channelDataStoreKeySet) {
+		if (!chainDataStoreKeySet.has(storeKey)) {
+			throw new Error(
+				`Channel data store key ${storeKey.toString('hex')} is missing in chain data store.`,
+			);
+		}
+	}
+
+	for (const storeKey of chainValidatorsStoreKeySet) {
+		if (!chainDataStoreKeySet.has(storeKey)) {
+			throw new Error(
+				`Chain validators store key ${storeKey.toString('hex')} is missing in chain data store.`,
+			);
+		}
+	}
```

Comment:
For each entry in the chain data substore there must be a corresponding entry (with the same value of storeKey) in the outbox root, channel data, and chain validators substores and viceversa